src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -419,6 +419,7 @@ private boolean run() {
                     t.fireCompletionHandlers(o); // do this after ErrorAction is set above
 
                     threads.remove(t.id);
+                    t.cleanUp();
```

Comment:
I think I like this better than #367 because it seems like there's fewer paths in which this call could be missed but I might not be understanding.